sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -191,9 +194,35 @@ object AnsiTypeCoercion extends TypeCoercionBase {
       case (DateType, TimestampType) => Some(TimestampType)
 
       // When we reach here, input type is not acceptable for any types in this type collection,
-      // try to find the first one we can implicitly cast.
+      // first try to find the all the expected types we can implicitly cast:
+      //   1. if there is no convertible data types, return None;
+      //   2. if there is only one convertible data type, cast input as it;
+      //   3. otherwise if there are multiple convertible data types, find the closet convertible
+      //      data type among them. If there is no such a data type, return None.
       case (_, TypeCollection(types)) =>
-        types.flatMap(implicitCast(inType, _, isInputFoldable)).headOption
+        // Since Spark contains special objects like `NumericType` and `DecimalType`, which accepts
+        // multiple types and they are `AbstractDataType` instead of `DataType`, here we use the
+        // conversion result their representation.
+        val convertibleTypes = types.flatMap(implicitCast(inType, _, isInputFoldable))
+        if (convertibleTypes.isEmpty) {
+          None
+        } else {
+          // find the closet convertible data type, which can be implicit cast to all other
+          // convertible types.
+          val closestConvertibleType = convertibleTypes.find { dt =>
```

Comment:
How about doing it like this so that we could quickly notice potential bugs?
```
          val foundTypes = convertibleTypes.filter { dt =>
            convertibleTypes.forall { target =>
              implicitCast(dt, target, isInputFoldable = false).isDefined
            }
          }

          assert(foundTypes.length <= 1)

          val closestConvertibleType = foundTypes.headOption
```

Reply:
Actually, this logic always selects the convertible leftmost data type from the type precedence list ?

Reply:
> Actually, this logic always selects the convertible leftmost data type from the type precedence list ?

Yes. The pain point is NullType and String Literal has multiple type precedence lists.


![image](https://user-images.githubusercontent.com/1097932/111567348-0aa10d80-875c-11eb-9655-b4545dffc400.png)

Reply:
> How about doing it like this so that we could quickly notice potential bugs?

Here is my concern: if there are duplicated types(E.g. UDT) in `TypeCollection`, then the assert will fail.

Reply:
> Yes. The pain point is NullType and String Literal has multiple type precedence lists.

Ah, I see. For example, what if we have `from=NullType, to=TypeCollection[StringType, BinaryType]`? It seems `convertibleTypes ` is not empty, but `closestConvertibleType` is `None`?

Reply:
Yes, the `convertibleTypes` is [StringType, BinaryType]. Since there is no such a type in [StringType, BinaryType] that can be converted to all the other types, the `closestConvertibleType` is None.

Reply:
Ah, okay. I got it. Thanks for the explanation. If so, could you add more test for these negative cases (the cases where the logic returns `None`), too?

Reply:
Sure :)